### DISCUSSION TITLE
Wasm simd128 support

### DISCUSSION THREAD

**Discussion by myfreeer** [recp/cglm#282] [Ideas]
Based on the [compact code](https://github.com/emscripten-core/emscripten/blob/3.1.34/system/include/compat/xmmintrin.h) by [emscripten](https://github.com/emscripten-core/emscripten), most part of the x86 [sse2](https://github.com/recp/cglm/tree/master/include/cglm/simd/sse2) simd code can be rewritten into wasm simd128.
Initial work [here](https://github.com/myfreeer/cglm), it should compile both library and tests with clang and [wasi-sdk](https://github.com/WebAssembly/wasi-sdk) with cmake arguments below:
```
-DCMAKE_C_FLAGS="-msimd128 -D_WASI_EMULATED_PROCESS_CLOCKS"
-DCMAKE_EXE_LINKER_FLAGS="-lwasi-emulated-process-clocks"
-DCGLM_STATIC=ON
-DCMAKE_TOOLCHAIN_FILE=/path/to/wasi-sdk-19.0/share/cmake/wasi-sdk.cmake
-DWASI_SDK_PREFIX=/path/to/wasi-sdk-19.0
-DCGLM_USE_TEST=ON
```
Where `/path/to/wasi-sdk-19.0/` is the path to extracted wasi sdk.

Here is an example of compiled tests: 
[compiled_tests_for_wasi.zip](https://github.com/recp/cglm/files/11010208/compiled_tests_for_wasi.zip)
This should run on [wasmtime](https://github.com/bytecodealliance/wasmtime) and [wasmer](https://github.com/wasmerio/wasmer) runtime with simd128 support and pass the tests.
![image](https://user-images.githubusercontent.com/17702502/226153077-1c95f7f1-2166-4bff-9db6-dec9b92a964d.png)

**Comment by recp** [comment]
Hi @myfreeer,

Wow that's great! Would you like to bring this via a PR?

**Reply by myfreeer** [reply]
There are still things to do, imcomplete listed below:
1. Inline or rewrite sse2 wrappers borrowed from emscripten
2. Change in cmake scripts for cmake arguments, or a build guide
3. Code style changes

So maybe a draft pr could be made first, and it would take some time to finish this.

**Reply by recp** [reply]
@myfreeer many thanks, I'll wait that PR :)

**Comment by recp** [comment]
@myfreeer is it possible to get additional **wasm** improvements and postponed tasks at https://github.com/recp/cglm/pull/286 :) moving **Compiling** section (with or without more documentation) in the PR to README, build.rst could be nice too

**Reply by myfreeer** [reply]
It's here: https://github.com/recp/cglm/pull/295

**Reply by recp** [reply]
@myfreeer many thanks

**Comment by kzhsw** [comment]
It seems that the current wasm compiler lacks compile-time contant evaluation feature, here is an example taken from `simd/x86/quat.h` and `simd/wasm/quat.h`:

Wasm:
shuffle + store
![screenshot wasm](https://github.com/recp/cglm/assets/120431159/6cb15825-0831-4b70-913d-a082a3c1d365)

x86:
movaps + movups
![screenshot x86](https://github.com/recp/cglm/assets/120431159/fd659d85-25f7-4005-ac4e-37dc3b4bbf8e)

Wasm link: <https://godbolt.org/z/aoEer7cd9>
x86 link: <https://godbolt.org/z/aoEer7cd9>

**Reply by myfreeer** [reply]
The link for x86 is the same as wasm, here is a reproduced x86 link with code above: <https://godbolt.org/z/qoWeEEnan>, maybe this should be reported to [llvm-project](https://github.com/llvm/llvm-project) instead.

**Reply by myfreeer** [reply]
It does not seem to reproduce on `Debian clang version 14.0.6` and `clang version 16.0.0` (wasi-sdk 20), I manually compiled this to wasm, but [wasm2wat](https://webassembly.github.io/wabt/demo/wasm2wat/) shows a different result.
![image](https://github.com/recp/cglm/assets/17702502/3f45857b-6e11-4a21-82b1-53292f4f21b5)
[source_and_compiled_binary_attached_here.zip](https://github.com/recp/cglm/files/12207058/source_and_compiled_binary_attached_here.zip)

**Reply by gottfriedleibniz** [reply]
One potential alternative would be to experiment with `wasm_i32x4_shuffle` versus `__builtin_shufflevector`. The former is emitting `@llvm.wasm.shuffle` and there may be some baggage/boundary/something preventing its evaluation.

**Reply by kzhsw** [reply]
It seems that `__builtin_shufflevector` is more effecient in godbolt clang here <https://godbolt.org/z/Yx5vY9ndc>, futher tests might also be needed in local compilers like wasi clang.
![image](https://github.com/recp/cglm/assets/120431159/24ddbda3-c63d-46a0-ba41-8ec8a46471de)

**Reply by recp** [reply]
Maybe we can use `__builtin_shufflevector` instead of `wasm_i32x4_shuffle` when possible ( maybe some other SIMD instructions too ), because it seems `wasm_i32x4_shuffle` is not well-supported: https://emscripten.org/docs/porting/simd.html?highlight=exception, something like **CGLM_WASM_ENABLE_EXPERIMENT** could force to use `wasm_i32x4_shuffle` instead of `__builtin_shufflevector` ...?

**Comment by myfreeer** [comment]
Found a type [here](https://github.com/recp/cglm/releases/tag/v0.9.0)
![image](https://github.com/recp/cglm/assets/17702502/2c21bf43-35b3-47d6-adf8-a66d025d0038)
